### PR TITLE
Delete user k8s, etc. certificates on re-issue

### PR DESCRIPTION
**Pull request by tcsc**
Prior to this change, user certificates for services like kubernetes were
preserved across a certificate re-issue. This led to issues where elevated
privileges granted by an access request were not applied to the service
certificates as they were not updated during the reissue process.

This patch changes the certificate re-issue process such that:
 * certificates for services (like Kuberenetes) are not preserved
   over a certificate re-issue. It is expected that they will be recreated
   on the first access to the service in question, and
 * the local keystore files for these certificates services are explicitly
   deleted so that the now-invalid cached certificates are not returned
   on keystore queries.

See-Also: #5047

### CODE REVIEW

File: lib/client/client.go [hunk 1]
```diff
@@ -186,8 +186,22 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
...
+	localKeyAgent := proxy.localAgent()
+
+	// The re-issued certificate may contain new roles that change our
+	// permissions with *any* service (e.g. Kubernetes, Database) in the
+	// cluster. We want to make sure no cached certificates are preserved
+	// once we have the re-issued certificate, forcing each service to get
+	// a new certificate (with the potentially updated roles) on first
+	// access, so we explicitly delete them all.
+	log.Debugf("Deleting all user certs...")
+	if err = localKeyAgent.DeleteUserCerts(params.RouteToCluster, WithAllCerts...); err != nil {
```
Comment (andrejtokarcik) [inline lib/client/client.go:198]: Wouldn't this effectively prevent more than a single user/service cert to be stored per cluster at a time? It seems this deletion should occur only if the list of active requests differs from the one used with the stored key.

Also, given the list of active requests should be obtainable from the stored key itself, this logic could be more appropriately moved directly into `LocalKeyAgent.AddKey`.

File: lib/client/client.go [hunk 2]
```diff
@@ -186,8 +186,22 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
...
+		log.Errorf("Deleting user certs failed")
```
Comment (andrejtokarcik) [inline lib/client/client.go:199]: Assuming writing these logs is really needed here:
```suggestion
		log.WithError(err).Errorf("Deleting user certs failed.")
```

File: lib/client/client.go [hunk 3]
```diff
@@ -198,7 +212,13 @@ func (proxy *ProxyClient) reissueUserCerts(ctx context.Context, params ReissuePa
...
 	if key == nil {
 		var err error
-		key, err = proxy.localAgent().GetKey(params.RouteToCluster, WithAllCerts...)
+		// Note that we are explicitly NOT persisting any existing service
+		// (e.g. DB, Kubernetes, etc) certificates across the re-issue of the
+		// main certificate. This is because the re-issued certificate may
+		// contain new roles, and preserving the old certificates will
+		// only stop those new roles being applied to the service
+		// certificate.
+		key, err = proxy.localAgent().GetKey(params.RouteToCluster)
```
Comment (andrejtokarcik) [inline lib/client/client.go:221]: Just to avoid any misunderstanding: This call has nothing to do with *persisting* the key data in the store itself. It's meant just for the convenience of the caller in case they happen to make use of k8s clusters/DBs/apps not mentioned in the reissue request itself.

File: lib/auth/permissions.go [hunk 4]
```diff
@@ -578,13 +578,13 @@ func contextForBuiltinRole(r BuiltinRole, clusterConfig services.ClusterConfig)
 	}, nil
 }
 
-func contextForLocalUser(u LocalUser, identity services.UserGetter, access services.Access) (*Context, error) {
+func contextForLocalUser(u LocalUser, identitySvc services.UserGetter, access services.Access) (*Context, error) {
```
Comment (awly) [inline lib/auth/permissions.go:581]: this change seems unrelated to the PR

File: lib/client/client.go (hunk 1 above)
Comment (awly) [reply lib/client/client.go:198]: +1, existing cert deletion should happen only when requesting additional roles, not on every reissue
please move this to `/tool/tsh/`, right after a successful access request approval

Comment (tcsc) [reply lib/client/client.go]: Moved.

File: lib/client/client.go (hunk 3 above)
Comment (tcsc) [reply lib/client/client.go]: Removed during refactor

File: lib/auth/permissions.go (hunk 4 above)
Comment (tcsc) [reply lib/auth/permissions.go]: Reverted

File: tool/tsh/tsh.go [hunk 5]
```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
...
+
+	oldKey, err := tc.LocalAgent().GetKey(profile.Cluster, client.WithSSHCerts{})
+	switch {
+	case trace.IsNotFound(err):
+		oldKey = nil
+	case err != nil:
 		return trace.Wrap(err)
+	default:
+		// carry on
 	}
```
Comment (andrejtokarcik) [inline tool/tsh/tsh.go:2048]:
```suggestion
	if err != nil && !trace.IsNotFound(err) {
		return trace.Wrap(err)
	}
```

File: tool/tsh/access_request_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+			expectChange: true,
+		}, {
+			desc:         "Duplicate insertion",
+			rolesOld:     []string{"alpha", "beta", "gamma"},
+			rolesNew:     []string{"alpha", "alpha", "gamma"},
+			expectChange: true,
+		},
+	}
+	keyGen := newKeygen(len(testCases))
+	defer keyGen.Close()
```
Comment (andrejtokarcik) [inline tool/tsh/access_request_test.go:134]: Prefer `t.Cleanup` to `defer` in tests.

**Review by andrejtokarcik — CHANGES_REQUESTED** [review]
I'm worried the proposed logic is incorrect as it assumes the impact of a role change to be limited to a single cluster whereas I think a role change should discard certs cached for all clusters so that requests against any cluster would trigger a reissue with the new roles.

Further, since SSH certs are now also cached per cluster, they should get discarded with the other service certs as well.

File: tool/tsh/access_request_test.go (hunk 6 above)
Comment (tcsc) [reply tool/tsh/access_request_test.go:134]: Happy to oblige, but what's the rationale?

File: tool/tsh/access_request.go [hunk 7]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+	result := make([]client.CertOption, 0, len(client.WithAllCerts))
+	for _, opt := range client.WithAllCerts {
+		if _, ok := opt.(client.WithSSHCerts); ok {
+			continue
+		}
+
+		result = append(result, opt)
+	}
+	return result
+}
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go]: The reissue code expliicitly sets the SSH cert in the key before I get to see it, and deleting it has caused issues downstream. I will attempt to get a "delete everything" approach working, though.

File: tool/tsh/access_request.go [hunk 8]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+	clusterName string, keyAgent *client.LocalKeyAgent,
+	options ...client.CertOption) error {
+
+	if oldKey == nil {
+		// If there is no such key in the store, then we can't possibly
+		// need to clean up any certificates attached to it.
+		return nil
+	}
+
+	// If we get here, we have an old key that matches the new key's index. If
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go:345]: Added

File: tool/tsh/access_request.go [hunk 9]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+	}
+
+	if rolesChanged {
+		err = keyAgent.DeleteUserCerts(newKey.ClusterName, options...)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go:364]: Hmm. I can see your point, but I'm not 100% sold on conflating the "error" and "success but nothing more to do" cases.

File: tool/tsh/tsh.go (hunk 5 above)
Comment (tcsc) [reply tool/tsh/tsh.go:2048]: Never mind. I'm drawing the "old" Roles list from the profile now, so the whole thing disappears

**Comment by tcsc** [comment]
> Further, since SSH certs are now also cached per cluster, they should get discarded with the other service certs as well.

From wading around in `tsh`, the SSH certs appear to be special; they seem to be the ultimate source of a lot of profile information for `tsh` (e.g. Roles & Active Requests), and a lot of code depends on them just existing at all times. That is not to say that they _shouldn't_ be discarded with the other certs, just that it opens up a lot more complexity than one might think.

I'm not sure of the best way forward with this.

Maybe the correct thing is to stop trying to cobble together a solution out the existing key management primitives, and write something that:
1. Enumerates all the known clusters for the active user behind the given proxy 
2. Deletes the service certs for all of them
3. Generates (somehow!) a new SSH cert for the user/cluster

File: tool/tsh/tsh.go [hunk 10]
```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
...
 	if params.RouteToCluster == "" {
 		params.RouteToCluster = profile.Cluster
 	}
-	if err := tc.ReissueUserCerts(cf.Context, params); err != nil {
+
+	// Look up any existing key for the cluster. We will need this to check whether
+	// the granted roles have changed post-approval. For our purposes here, it's OK
+	// for this oldKey not to exist.
+
+	oldKey, err := tc.LocalAgent().GetKey(profile.Cluster, client.WithSSHCerts{})
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/tsh.go:2040]: I've changed the code to take the baseline roleset from the active profile, which is already loaded in this function anyway. This _seems_ like a better place, but please let me know if that's a bad assumption.

File: tool/tsh/tsh.go [hunk 11]
```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
...
+	// Check to see if there is a mismatch between the roles granted by the
+	// old key (if any) and the new  key, then delete the user certificates
+	// for the cluster if such a mismatch is detected.
+	//
+	// NB: The SSH cert will have been explicitly updated by the re-issue
+	//     process, so we don't want to delete that. All of the rest are
+	//     fair game.
+
+	err = deleteUserCertsOnRoleMismatch(oldKey, newKey, newKey.ClusterName,
+		tc.LocalAgent(), client.WithAllCertsExceptSSH...)
```
Comment (awly) [inline tool/tsh/tsh.go:2068]: As @andrejtokarcik mentioned, we have multiple SSH certs now, one for each cluster.
`newKey` will contain the new SSH cert.
How about deleting _all_ certs, including SSH ones, and then writing `newKey` again?
I believe `tc.ActivateKey(cf.Context, newKey)` should do that

File: tool/tsh/access_request.go [hunk 12]
```diff
@@ -289,3 +291,75 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+
+	oldRoleSet := make(map[string]int)
+	for _, r := range oldRoles {
+		oldRoleSet[r] = 0
+	}
+
+	newRoleSet := make(map[string]int)
+	for _, r := range newRoles {
+		newRoleSet[r] = 0
+	}
```
Comment (awly) [inline tool/tsh/access_request.go:317]:
```suggestion
	oldRoleSet := make(map[string]struct{})
	for _, r := range oldRoles {
		oldRoleSet[r] = struct{}{}
	}

	newRoleSet := make(map[string]struct{})
	for _, r := range newRoles {
		newRoleSet[r] = struct{}{}
	}
```

File: tool/tsh/access_request.go [hunk 13]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+// deleteUserCertsOnRoleMismatch() checks to see if there is a difference
+// between the roleset supplied by an old key and a new one. If a mismatch
+// exists, it will delete the user certs for the cluster
+//
+// Note that it's legitimate behaviour for `oldKey` to be nil; this just
+// implies that there are no "old" credentials for us to expire and there
+// is nothing to do.
+func deleteUserCertsOnRoleMismatch(oldKey, newKey *client.Key,
+	clusterName string, keyAgent *client.LocalKeyAgent,
+	options ...client.CertOption) error {
```
[the remark this answers is not in the document]
Comment (awly) [reply tool/tsh/access_request.go:331]: @andrejtokarcik why do you think it should be in `LocalKeyAgent`?

File: tool/tsh/access_request.go [hunk 14]
```diff
@@ -289,3 +291,75 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+	if oldKey == nil {
+		// If there is no such key in the store, then we can't possibly
+		// need to clean up any certificates attached to it.
+		return nil
+	}
+
+	if oldKey.KeyIndex != newKey.KeyIndex {
+		return trace.BadParameter(
+			"old key index %+v does not match new key index %+v",
+			oldKey.KeyIndex, newKey.KeyIndex)
```
Comment (awly) [inline tool/tsh/access_request.go:342]: nit: don't wrap long lines https://github.com/golang/go/wiki/CodeReviewComments#line-length

File: tool/tsh/access_request.go [hunk 15]
```diff
@@ -289,3 +291,75 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+	// reflect the permissions that the auth server is granting the user.
+	//
+	// Let's delete the current set of certificates so that they cannot be
+	// used by mistake
+	rolesChanged, err := rolesHaveChanged(oldKey, newKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if rolesChanged {
```
Comment (awly) [inline tool/tsh/access_request.go:357]: while it's nice to minimize disk IO, doing this check may be unnecessary
this function is only called during an access request flow, so user roles are most likely going to change

WDYT about removing all the roleset checking logic, and deleting the certs unconditionally?

File: tool/tsh/access_request_test.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+		Pub:  pub,
+		Cert: cert,
+	}
+
+	return key, nil
+}
+
+func TestPassingNilToIsNotFoundReturnsFalse(t *testing.T) {
+	require.False(t, trace.IsNotFound(nil))
+}
```
Comment (awly) [inline tool/tsh/access_request_test.go:80]: was this for debugging?

Comment (tcsc) [reply tool/tsh/access_request_test.go:80]: Just asserting something I wasn't sure of.

File: tool/tsh/tsh.go (hunk 11 above)
Comment (tcsc) [reply tool/tsh/tsh.go:2068]: I'll give it a shot and let you know.

File: tool/tsh/access_request.go (hunk 15 above)
Comment (tcsc) [reply tool/tsh/access_request.go:357]: That was the initial implementation, so yeah - I have no particular issues there. That would also be a bit more future-proof, should we start granting things _other_ than roles later on.

File: tool/tsh/access_request.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply tool/tsh/access_request.go:331]: Well, from my perspective the whole delete-on-roleset-change logic would ideally occur inside `LocalKeyAgent.AddKey`, similarly to delete-on-private-key-change that's already performed there.

My reasoning is that we should be defensive and not expect explicit role requests to be the only way in which the roleset may be updated (even if it currently is the case). The logic should be made generally available (and even triggered "for free") for new functionality that may change roles or even current use cases that we might've missed.

File: tool/tsh/access_request_test.go (hunk 16 above)
Comment (tcsc) [reply tool/tsh/access_request_test.go]: Deleted

File: tool/tsh/access_request.go (hunk 15 above)
Comment (tcsc) [reply tool/tsh/access_request.go]: Restored unconditional deletion

File: tool/tsh/access_request.go (hunk 14 above)
Comment (tcsc) [reply tool/tsh/access_request.go]: Deleted anyway

File: tool/tsh/access_request.go (hunk 12 above)
Comment (tcsc) [reply tool/tsh/access_request.go]: Deleted in favour of unconditionally deleting cert cache, but at least made me look up https://dave.cheney.net/2014/03/25/the-empty-struct

File: tool/tsh/tsh.go (hunk 11 above)
Comment (tcsc) [reply tool/tsh/tsh.go]: Yep, that's the new approach now. Seems to work in latest patch.

**Comment by tcsc** [comment]
@awly's suggestion to go back to unconditionally deleting the certificates without 
checking for a role change simplifies everything. This, hopefully last, patch makes 
the following changes
  
- Deletes all of the role change detection code
- Adds a parameter to `ProxyClient.ReissueUserCerts` that indicates how existing
  certificates in the cache should be treated (i.e. Kept or Dropped)
- Adds handling ProxyClient.ReissueUserCerts() that takes the appropriate cache action.
- Adds handling to ProxyClient.reissueUserCerts() that changes how cached certificates
  are preserved across a re-issue.
    
I've added this to the ProxyClient as the cache cleanup needs to happen _before_ the re-issued `Key` is added to the key store, otherwise it becomes very difficult to differentiate between the old certificates (that need purging) and the new certs (that need to be preserved).

Given that the rest of the system currently expects the new certs to be in the key store before `ProxyClient.ReissueUserCerts()` returns, this looked like the minimal change required to add the desired behaviour.

File: tool/tsh/access_request.go [hunk 17]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
 	fmt.Fprintf(os.Stderr, "\nhint: use 'tsh request show <request-id>' for additional details\n")
 	return trace.Wrap(err)
 }
+
+// rolesHaveChanged checks to see if there is a difference in the rolesets
+// applied to two certificates. Handles duplicates and different sort
+// orders in the rolesets.
+func rolesHaveChanged(oldKey, newKey *client.Key) (bool, error) {
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go]: Deleted in change of approach

File: tool/tsh/access_request.go [hunk 18]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+		return true, nil
+	}
+
+	// Sort the roles in each key so that we can assert that the contents
+	// of both rolesets is identical, allowing for duplicates and different
+	// ordering. There are probably faster ways to do this, but we'll worry
+	// about that if/when this starts to dominate the performance of `tsh`.
+
+	sort.Strings(oldRoles)
+	sort.Strings(newRoles)
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go]: Deleted in change of approach

File: tool/tsh/access_request.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go]: "Delete everything" works, it just needs to be inthe middle of `ReissueCertificates()` to work, otherwise the plumbing gets waaaaay too complicated.

File: tool/tsh/access_request.go [hunk 19]
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
...
+
+	for i := range oldRoles {
+		if oldRoles[i] != newRoles[i] {
+			return true, nil
+		}
+	}
+
+	return false, nil
```
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go]: Deleted during change of approach

File: tool/tsh/access_request.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/access_request.go]: Deleted in change of approach

File: tool/tsh/tsh.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tcsc) [reply tool/tsh/tsh.go]: Deleted in change of approach

File: tool/tsh/tsh.go (hunk 5 above)
Comment (tcsc) [reply tool/tsh/tsh.go]: And the whole think gas been deleted in  any case

**Review by awly — APPROVED** [review]

**Review by nklaassen — APPROVED** [review]
LGTM

File: lib/client/client.go (hunk 2 above)
Comment (tcsc) [reply lib/client/client.go]: Refactored out of existence

**Review by andrejtokarcik — APPROVED** [review]

**Review by andrejtokarcik — COMMENTED** [review]
BTW, does this really work as intended since it doesn't seem to be performing any real deletion of the stored certs anymore?

File: lib/client/client.go [hunk 20]
```diff
@@ -178,27 +178,58 @@ func (p ReissueParams) isMFARequiredRequest(sshLogin string) *proto.IsMFARequire
...
+
+const (
+	// CertCacheDrop indicates that all user certificates should be dropped as
+	// part of the re-issue process. This can be necessary if the roles
+	// assigned to the user are expected to change as a part of the re-issue.
+	CertCacheDrop CertCachePolicy = 0
+
+	// CertCacheKeep indicates that all user certificates (except those
+	// explicitly updated by the re-issue) should be preserved across the
+	// re-issue process
```
Comment (andrejtokarcik) [inline lib/client/client.go:193]:
```suggestion
	// re-issue process.
```

File: lib/client/client.go [hunk 21]
```diff
@@ -178,27 +178,58 @@ func (p ReissueParams) isMFARequiredRequest(sshLogin string) *proto.IsMFARequire
...
-func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissueParams) error {
-	key, err := proxy.reissueUserCerts(ctx, params)
+func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, cachePolicy CertCachePolicy, params ReissueParams) error {
+	key, err := proxy.reissueUserCerts(ctx, cachePolicy, params)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
+	if cachePolicy == CertCacheDrop {
+		proxy.localAgent().DeleteUserCerts("", WithAllCerts...)
```
Comment (andrejtokarcik) [inline lib/client/client.go:206]: Missed this, sorry, ignore my question above.

**Review by russjones — APPROVED** [review]
Bot.
